File: src/features/chat/messages/messages.go
// Package messages owns the chat message markup. Each message kind is a
// named template in messages.html, emitted at two points from the one
// definition: filled in with real data for server-rendered history, and
// zero-valued inside <template> tags for the client to clone while streaming.
package messages

import (
	"bytes"
	"ct-go-chat/src/components/component"
	"ct-go-chat/src/infrastructure/agent"
	"ct-go-chat/src/infrastructure/agent/bedrock"
	_ "embed"
	"encoding/json"
	"fmt"
	"html/template"
)

//go:embed messages.html
var messagesHTML string

//go:embed messages.js
var messagesJS string

// The same source is parsed twice: templatesComp renders the root <template>
// block (plus script) for the client, partials executes the named message
// templates with real data for history.
var templatesComp = component.WithIIFE("messages.html", messagesHTML, messagesJS)
var partials = template.Must(template.New("messages.html").Parse(messagesHTML))

type userProps struct {
	Text string
}

type assistantProps struct {
	SegmentsHTML template.HTML
	DetailsHTML  template.HTML
	// HideToolbar drops the copy toolbar entirely (cancelled exchanges).
	HideToolbar bool
	// Final marks a finished exchange: toolbar visible, aria-live polite.
	// The zero value is the live-streaming shell the client clones.
	Final bool
}

type textProps struct {
	Text string
}

type toolProps struct {
	Name    string
	Input   string
	Output  string
	IsError bool
}

type detailsTriggerProps struct {
	PopoverTarget string
	OutputTokens  string
}

type detailsProps struct {
	ID               string
	InputTokens      string
	CacheWriteTokens string
	CacheReadTokens  string
	OutputTokens     string
	Cost             string
	TTFB             string
	TTLB             string
}

type templateSlots struct {
	User           userProps
	Assistant      assistantProps
	Text           textProps
	Tool           toolProps
	DetailsTrigger detailsTriggerProps
	Details        detailsProps
}

// RenderTemplates emits the <template> block and script the client uses to
// render messages while streaming.
func RenderTemplates() (template.HTML, error) {
	return templatesComp.Render(templateSlots{})
}

// RenderHistory renders a conversation's exchanges as message HTML. Assistant
// text is emitted as raw markdown text content; the client renders it on load.
func RenderHistory(exchanges []agent.Exchange) (template.HTML, error) {
	var out template.HTML
	for _, ex := range exchanges {
		user, err := renderPartial("message-user", userProps{Text: ex.Request})
		if err != nil {
			return "", err
		}
		assistant, err := renderAssistant(ex)
		if err != nil {
			return "", err
		}
		out += user + assistant
	}
	return out, nil
}

func renderAssistant(ex agent.Exchange) (template.HTML, error) {
	segments, err := renderSegments(ex.Rounds)
	if err != nil {
		return "", err
	}

	if ex.Cancelled {
		cancelled, err := renderPartial("message-cancelled", nil)
		if err != nil {
			return "", err
		}
		if segments == "" {
			return cancelled, nil
		}
		assistant, err := renderPartial("message-assistant", assistantProps{
			SegmentsHTML: segments,
			HideToolbar:  true,
		})
		if err != nil {
			return "", err
		}
		return assistant + cancelled, nil
	}

	details, err := renderDetails(ex)
	if err != nil {
		return "", err
	}
	return renderPartial("message-assistant", assistantProps{
		SegmentsHTML: segments,
		DetailsHTML:  details,
		Final:        true,
	})
}

func renderSegments(rounds []agent.Round) (template.HTML, error) {
	var out template.HTML
	for _, round := range rounds {
		if round.Text != "" {
			seg, err := renderPartial("message-assistant-text", textProps{Text: round.Text})
			if err != nil {
				return "", err
			}
			out += seg
		}
		for _, call := range round.ToolCalls {
			seg, err := renderPartial("message-tool", toolProps{
				Name:    call.Name,
				Input:   prettyJSON(call.Input),
				Output:  call.Result,
				IsError: call.IsError,
			})
			if err != nil {
				return "", err
			}
			out += seg
		}
	}
	return out, nil
}

func renderDetails(ex agent.Exchange) (template.HTML, error) {
	if ex.ID == "" {
		return "", nil
	}
	trigger, err := renderPartial("message-details-trigger", detailsTriggerProps{
		PopoverTarget: ex.ID,
		OutputTokens:  bedrock.DisplayInt(ex.Usage.OutputTokens) + " tok",
	})
	if err != nil {
		return "", err
	}
	details, err := renderPartial("message-details", detailsProps{
		ID:               ex.ID,
		InputTokens:      bedrock.DisplayInt(ex.Usage.InputTokens),
		CacheWriteTokens: bedrock.DisplayInt(ex.Usage.CacheCreationInputTokens),
		CacheReadTokens:  bedrock.DisplayInt(ex.Usage.CacheReadInputTokens),
		OutputTokens:     bedrock.DisplayInt(ex.Usage.OutputTokens),
		Cost:             bedrock.DisplayCost(ex.Usage.CostUSD),
		TTFB:             fmt.Sprintf("%d ms", ex.Timing.TTFBMs),
		TTLB:             fmt.Sprintf("%d ms", ex.Timing.TTLBMs),
	})
	if err != nil {
		return "", err
	}
	return trigger + details, nil
}

func renderPartial(name string, data any) (template.HTML, error) {
	var buf bytes.Buffer
	if err := partials.ExecuteTemplate(&buf, name, data); err != nil {
		return "", fmt.Errorf("messages: render %s: %w", name, err)
	}
	return template.HTML(buf.String()), nil
}

func prettyJSON(raw json.RawMessage) string {
	var buf bytes.Buffer
	if err := json.Indent(&buf, raw, "", "  "); err != nil {
		return string(raw)
	}
	return buf.String()
}

File: src/features/chat/messages/messages.js
// marked and DOMPurify load deferred, so the renderer is built on first use
// rather than at parse time of this inline script.
let renderMarkdown = null;

function markdownRenderer(rawText) {
  if (!renderMarkdown) {
    const renderer = new marked.Renderer();
    renderer.heading = ({ tokens, depth }) => {
      const text = renderer.parser.parseInline(tokens);
      const level = Math.min(depth + 1, 6);
      return `<h${level}>${text}</h${level}>`;
    };
    marked.setOptions({ gfm: true, breaks: false });
    renderMarkdown = (raw) =>
      DOMPurify.sanitize(marked.parse(raw, { renderer }));
  }
  return renderMarkdown(rawText);
}

function scrollToBottom() {
  window.scrollTo({ top: document.body.scrollHeight, behavior: "smooth" });
}

function templateContent(id) {
  return document.getElementById(id).content.cloneNode(true).firstElementChild;
}

function cloneTemplate(id, messages, before) {
  const node = templateContent(id);
  messages.insertBefore(node, before);
  scrollToBottom();
  return node;
}

function copyWithFeedback(button, text) {
  navigator.clipboard.writeText(text).then(() => {
    const icon = button.querySelector(".icon");
    icon.textContent = "check";
    setTimeout(() => (icon.textContent = "content_copy"), 2500);
  });
}

function appendUserMessage(text, messages, before) {
  const node = cloneTemplate("message-user", messages, before);
  node.querySelector(".message-text").textContent = text;
}

function appendMessageUsage(messageNode, exchange) {
  if (!exchange?.usage || !exchange.id) return;
  const { usage } = exchange;
  const popoverTrigger = templateContent("message-details-trigger");
  const popover = templateContent("message-details");
  popover.setAttribute("id", exchange.id);
  popover.querySelector(".input-tokens").textContent =
    usage.display.input_tokens;
  popover.querySelector(".cache-write-tokens").textContent =
    usage.display.cache_creation_input_tokens;
  popover.querySelector(".cache-read-tokens").textContent =
    usage.display.cache_read_input_tokens;
  popover.querySelector(".output-tokens").textContent =
    usage.display.output_tokens;
  popover.querySelector(".cost").textContent = usage.display.cost;
  if (exchange.timing) {
    popover.querySelector(".ttfb").textContent = exchange.timing.ttfb_ms + " ms";
    popover.querySelector(".ttlb").textContent = exchange.timing.ttlb_ms + " ms";
  }
  popoverTrigger.querySelector(".output-tokens").textContent =
    `${usage.display.output_tokens} tok`;
  popoverTrigger.setAttribute("popovertarget", exchange.id);

  messageNode.querySelector(".message-copy").after(popoverTrigger);
  popoverTrigger.after(popover);
}

function appendAssistantMessage(messages, before) {
  const node = cloneTemplate("message-assistant", messages, before);
  const segments = node.querySelector(".message-segments");
  let rawText = "";
  let segmentText = "";
  let segmentElement = null;
  let openToolElement = null;
  let renderTimer = null;

  function render() {
    if (segmentElement) {
      segmentElement.innerHTML = markdownRenderer(segmentText);
    }
  }

  function flushRender() {
    if (renderTimer) {
      clearTimeout(renderTimer);
      renderTimer = null;
    }
    render();
  }

  return {
    appendText(text) {
      if (!segmentElement) {
        const segment = templateContent("message-assistant-text");
        segmentElement = segment.querySelector(".message-text");
        segments.appendChild(segment);
        segmentText = "";
        if (rawText) rawText += "\n\n";
      }
      segmentText += text;
      rawText += text;
      if (renderTimer) return;
      renderTimer = setTimeout(() => {
        renderTimer = null;
        render();
      }, 60);
      scrollToBottom();
    },
    appendTool(tool) {
      flushRender();
      segmentElement = null;

      openToolElement = templateContent("message-tool");
      openToolElement.querySelector(".tool-name").textContent = tool.name;
      openToolElement.querySelector(".tool-input").textContent =
        JSON.stringify(tool.input, null, 2);
      segments.appendChild(openToolElement);
      scrollToBottom();
    },
    toolResult(tool) {
      if (!openToolElement) return;
      const output = openToolElement.querySelector(".tool-output");
      output.textContent = tool.result;
      if (tool.is_error) {
        output.classList.add("text-error");
      }
      openToolElement = null;
      scrollToBottom();
    },
    finalise(exchange) {
      flushRender();

      node.setAttribute("aria-live", "polite");
      node
        .querySelector(".message-copy")
        .addEventListener("click", (e) =>
          copyWithFeedback(e.currentTarget, rawText),
        );
      appendMessageUsage(node, exchange);
      node.querySelector(".message-toolbar").classList.remove("opacity-0");
    },
    cancel() {
      flushRender();

      const anchor = node.nextSibling;
      if (segments.childElementCount > 0) {
        node.querySelector(".assistant-badge")?.classList?.remove("loading");
      } else {
        node.remove();
      }
      node.querySelector(".message-toolbar").remove();
      cloneTemplate("message-cancelled", messages, anchor);
    },
  };
}

// Server-rendered history arrives with raw markdown as text content. This
// one-shot pass renders it through the same markdown pipeline the live
// stream uses, wires the copy buttons, and jumps to the latest message.
document.addEventListener("DOMContentLoaded", () => {
  const messages = document.querySelector('[x-ref="messages"]');
  if (!messages) return;
  const hydrated = messages.querySelectorAll('[aria-label="Assistant"]');
  for (const node of hydrated) {
    const textElements = [
      ...node.querySelectorAll(".message-segments .message-text"),
    ];
    const rawText = textElements.map((el) => el.textContent).join("\n\n");
    for (const el of textElements) {
      el.innerHTML = markdownRenderer(el.textContent);
    }
    const copyButton = node.querySelector(".message-copy");
    if (copyButton) {
      copyButton.addEventListener("click", (e) =>
        copyWithFeedback(e.currentTarget, rawText),
      );
    }
  }
  if (hydrated.length > 0) {
    window.scrollTo(0, document.body.scrollHeight);
  }
});

window.chatMessages = { appendUserMessage, appendAssistantMessage };
